addons/account/models/account_partial_reconcile.py

```diff
@@ -126,8 +126,61 @@ def unlink(self):
             } for move in moves_to_reverse]
             moves_to_reverse._reverse_moves(default_values_list, cancel=True)
 
+        self._update_matching_number(all_reconciled)
         return res
 
+    @api.model_create_multi
+    def create(self, vals_list):
+        partials = super().create(vals_list)
+        all_reconciled = (partials.debit_move_id + partials.credit_move_id)._all_reconciled_lines()
+        self._update_matching_number(all_reconciled)
+        return partials
+
+    @api.model
+    def _update_matching_number(self, aml_ids):
+        self.env['account.partial.reconcile'].flush_model()
+        aml_ids.flush_recordset(['full_reconcile_id'])
+        self.env.cr.execute("""
+            WITH RECURSIVE partials (line_id, partial_id, direction) AS (
+                    SELECT line.id,
+                           COALESCE(credit.id, debit.id),
+                           CASE WHEN debit.id IS NOT NULL THEN 'debit' ELSE 'credit' END
+                      FROM account_move_line line
+                 LEFT JOIN account_partial_reconcile debit ON debit.credit_move_id = line.id
+                 LEFT JOIN account_partial_reconcile credit ON credit.debit_move_id = line.id
+                     WHERE COALESCE(credit.id, debit.id) IS NOT NULL
+                       AND line.full_reconcile_id IS NULL
+                       AND line.id = ANY(%s)
+
+                                                UNION
+
+                    SELECT p.line_id,
+                           CASE WHEN direction = 'debit' THEN debit.id ELSE credit.id END,
+                           CASE WHEN direction = 'debit' THEN 'credit' ELSE 'debit' END
+                      FROM partials p
+                 LEFT JOIN account_partial_reconcile this ON this.id = partial_id
+                 LEFT JOIN account_partial_reconcile debit ON this.debit_move_id = debit.debit_move_id
+                 LEFT JOIN account_partial_reconcile credit ON this.credit_move_id = credit.credit_move_id
+                     WHERE CASE WHEN direction = 'debit' THEN debit.id ELSE credit.id END != this.id
+            )
+            SELECT line_id, 'P' || MIN(partial_id)::text AS matching_number
+              FROM partials
+             GROUP BY line_id
```

Comment:
Turns out one of my first versions of the query runs faster than the current one on a real database :shrug: 

```suggestion
            WITH RECURSIVE partials (line_id, current_id) AS (
                    SELECT id, id FROM account_move_line

                                                UNION

                    SELECT p.line_id, 
                           CASE WHEN partial.debit_move_id = p.current_id THEN partial.credit_move_id 
                                ELSE partial.debit_move_id 
                           END
                      FROM partials p
                      JOIN account_partial_reconcile partial ON p.current_id = partial.debit_move_id 
                                                             OR p.current_id = partial.credit_move_id
            )
              SELECT line_id, MIN(partial.id) AS matching_number
                FROM partials
                JOIN account_partial_reconcile partial ON current_id = partial.debit_move_id 
                                                       OR current_id = partial.credit_move_id
            GROUP BY line_id
```

Should change it as well for the upgrade